circle/src/domain.rs

```diff
@@ -0,0 +1,355 @@
+use alloc::vec;
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+use p3_commit::{LagrangeSelectors, PolynomialSpace};
+use p3_field::extension::{Complex, ComplexExtendable};
+use p3_field::{batch_multiplicative_inverse, AbstractField, ExtensionField};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::Matrix;
+use p3_util::{log2_ceil_usize, log2_strict_usize};
+use tracing::instrument;
+
+use crate::util::{point_to_univariate, s_p_at_p, univariate_to_point, v_0, v_n};
+
+/// X is generator, O is the first coset, goes counterclockwise
+/// ```text
+///   O X .
+///  .     .
+/// .       O <- start = shift
+/// .   .   - (1,0)
+/// O       .
+///  .     .
+///   . . O
+/// ```
+///
+/// For ordering reasons, the other half will start at gen / shift:
+/// ```text
+///   . X O  <- start = gen/shift
+///  .     .
+/// O       .
+/// .   .   - (1,0)
+/// .       O
+///  .     .
+///   O . .
+/// ```
+///
+/// The full domain is the interleaving of these two cosets
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct CircleDomain<F> {
+    // log_n corresponds to the log size of the WHOLE domain
+    pub(crate) log_n: usize,
+    pub(crate) shift: Complex<F>,
+}
+
+impl<F: ComplexExtendable> CircleDomain<F> {
+    pub(crate) fn new(log_n: usize, shift: Complex<F>) -> Self {
+        Self { log_n, shift }
+    }
+    pub(crate) fn standard(log_n: usize) -> Self {
+        Self {
+            log_n,
+            shift: F::circle_two_adic_generator(log_n + 1),
+        }
+    }
+    fn is_standard(&self) -> bool {
+        self.shift == F::circle_two_adic_generator(self.log_n + 1)
+    }
+    pub(crate) fn points(&self) -> impl Iterator<Item = Complex<F>> {
+        let half_gen = F::circle_two_adic_generator(self.log_n - 1);
+        let coset0 = half_gen.shifted_powers(self.shift);
+        let coset1 = half_gen.shifted_powers(half_gen / self.shift);
+        coset0.interleave(coset1).take(1 << self.log_n)
+    }
+
+    #[instrument(skip_all, fields(log_n = %self.log_n))]
+    pub(crate) fn lagrange_basis<EF: ExtensionField<F>>(&self, point: Complex<EF>) -> Vec<EF> {
+        let domain = self.points().collect_vec();
+
+        // the denominator so that the lagrange basis is normalized to 1
+        // TODO: this depends only on domain, so should be precomputed
```

Comment:
Note to self: seems like this `s_p_at_p` alternates two values over the whole domain. Don't need to precompute.

Reply:
This is right. The normalization value at a single point of the circle domain is good enough to conclude by "symmetry" the value at the other points.